<!DOCTYPE html>
<html>
  <head>
    <title>Visualizing Genomics Data</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Visualizing Genomics Data
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/RU_VisualizingGenomicsData/" class="uri">http://rockefelleruniversity.github.io/RU_VisualizingGenomicsData/</a>

---




##The Course 

 
 
* The Course 
* [Importance of Visualising Genomics Data](#/vizdata). 
* [Reminder of file types](#/filetypes) 
* [Reminder of data types](#/datatypes) 
* [Materials](#/materials) 
* [Visualising genomics data in R](#/VizinR) 
* [Plotting genome axis](#/genomeaxis) 
* [Plotting genome data](#/datatracks) 
* [Plotting genome annotation](#/Annotation) 
* [Plotting genome sequence](#/seqtrack) 
* [Plotting genomic alignments](#/genomeaxis) 
* [Plotting from external databases](#/externaldata) 

---
##Importance of Visualising Genomics Data. 

It is an essential step in genomics data analysis to visualise your data. This allows you to review data for both known or unexpected data characteristics and potential artefacts. 
 
While we have discussed using IGV to review genomics data, now we will discuss how to do this while still working with in the R. 

---
##Visualising Genomics Data in R/Bioconductor. 

 
In complement to our [IGV genome browser course](https://rockefelleruniversity.github.io/IGV_course/) where we reviewed visualising genomics data in a browser, here we will use R/Bioconductor to produce publication quality graphics programatically.  
 
Much of the material will require some familiarity with R and Bioconductor [(you can revisit our courses on those here)](https://rockefelleruniversity.github.io/) and these will be used in tight conjunction with tools introduced today such as the Bioconductor package, **Gviz**. 
---
##Reminder of file types 

In this session we will be dealing with a range of data types. For more information on file types you can revisit our material. 
 
* [File Formats](https://rockefelleruniversity.github.io/Genomic_Data/). 
 
For more information on visualising genomics data in browsers you can visit our IGV course. 
 
* [IGV](https://rockefelleruniversity.github.io/IGV_course/). 
---
##Reminder of data types in Bioconductor 

We will also encounter and make use of many data structures and data types which we have seen throughout our courses on HTS data. You can revisit this material to refresh on HTS data analysis in Bioconductor and R below. 
 
* [Bioconductor](https://rockefelleruniversity.github.io/Bioconductor_Introduction/). 
* [Genomic Intervals](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicIntervals_In_Bioconductor.html)
* [Genomic Scores](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html)
* [Sequences](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/SequencesInBioconductor.html)
* [Gene Models](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicFeatures_In_Bioconductor.html)
* [Alignments](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignedDataInBioconductor.html). 
* [ChIP-seq](http://rockefelleruniversity.github.io/RU_ChIPseq/). 
* [ATAC-seq](http://rockefelleruniversity.github.io/RU_ATACseq/). 
* [RNA-seq](http://rockefelleruniversity.github.io/RU_RNAseq/). 
 
 
 
---
##Materials. 

All material for this course can be found on github. 
* [Visualising Genomics Data](https://github.com/RockefellerUniversity/RU_VisualizingGenomicsData) 
 
Or can be downloaded as a zip archive from here.  
* [Download zip](https://github.com/RockefellerUniversity/RU_VisualizingGenomicsData/archive/master.zip) 

---
##Materials. - Presentations, source code and practicals. 

Once the zip file in unarchived. All presentations as HTML slides and pages, their R code and HTML practical sheets will be available in the directories underneath. 
 
* **viz_course/presentations/Slides/** 
Presentations as an HTML slide show. 
* **viz_course/presentations/exercises/** 
Some tasks/examples to work through.

---
##Materials. - Data for presentations, practicals. 

 
 
All data to run code in the presentations and in the practicals is available in the zip archive. This includes coverage as bigWig files, aligned reads as BAM files and genomic intervals stored as BED files. 
 
We also include some RData files containing precompiled results from querying database (in case of external server downtime).

All data can be found under the **Data** directory 
 
**Data/** 
 
---
##Set the Working directory 

 
 
Before running any of the code in the practicals or slides we need to set the working directory to the folder we unarchived.  
 
You may navigate to the unarchived VisualisingGenomicsData folder in the Rstudio menu 
 
**Session -&gt; Set Working Directory -&gt; Choose Directory** 
 
or in the console. 
 

```r
setwd("/PathToMyDownload/VisualizingGenomicsData/viz_course/presentations/Slides") 
# e.g. setwd("~/Downloads/VisualizingGenomicsData/viz_course/presentations/Slides") 
```
 
 
 
---
##Why are we here? 

 
 
Genomics data can often be visualised in genome browsers such as the user friendly IGV genome browser. 
 
This allows for the visualisation of our processed data in its genomic context. 
 
In Genomics *(and most likely any Omics)*, it is important to review our data/results and hypotheses in a browser to identify patterns or potential artefacts discovered or missed within our analysis. 
---
##But we covered this?? 

 
 
We have already discussed on using the IGV browser to review our data and get access to online data repositories. 
 
IGV is quick, user friendly GUI to perform the essential task of review genomics data in its context. 
 
For more information see our course on IGV [here](http://mrccsc.github.io/IGV_course/). 
 
 
---
##Then why not just use IGV? 

 
 
Using a genome browser to review sites of interest across the genome is a critical **first** step. 
 
**Using processed and often indexed genomics data files**, IGV offers a method to rapidly interrogate genomics data along the linear genome. 
 
**IGV does its job well** and should always be an immediate early step in data review. By being good at this however it **does not offer the flexibility** in displaying data we wish to achieve, more so **when expecting to review a large number of sites**. 
---
##Visualising Genomics Data around Genomic Features in R (Gviz) 

The Gviz packages offers methods to produce publication quality plots of genomics data at genomic features of interest. 
 
 
To get started using Gviz in some biological examples, first we need to install the package. 
 

```r
## try http:// if https:// URLs are not supported 
source("https://bioconductor.org/biocLite.R") 
biocLite("Gviz") 
```
 
 
 
---
##Getting started with Gviz -- Linear genome axis. 

Gviz provides methods to plot many genomics data types (as with IGV) over genomic features and genomic annotation within a linear genomic reference. 
 
 
The first thing we can do then is set up our linear axis representing positions on genomes. 
 
For this we use our first function from **Gviz**, **GenomeAxisTrack()**. 
Here we use the **name** parameter to set the name to be "myAxis". 
 

```r
library(Gviz) 
```

```
## Loading required package: grid
```

```r
genomeAxis &lt;- GenomeAxisTrack(name="MyAxis") 
genomeAxis 
```

```
## Genome axis 'MyAxis'
```
---
##Getting started with Gviz -- Plotting the axis 

 
 
Now we have created a **GenomeAxisTrack** track object we can display the object using **plotTracks** function. 
 
In order to display a axis track we need to set the limits of the plot *(otherwise where would it start and end?)*. 
 

```r
plotTracks(genomeAxis,from=100,to=10100) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-5-1.png)&lt;!-- --&gt;
 
 
---
##Getting started with Gviz -- Configuring the axis (part-1) 

 
 
It is fairly straightforward to create and render this axis. 
Gviz offers a high degree of flexibility in the way these tracks can be plotted with some very useful plotting configurations included. 
 
A useful feature is to add some information on the direction of the linear genome represented in this **GenomeAxisTrack**. 
 
We can add labels for the 5' to 3' direction for the positive and negative strand by using the **add53** and **add35** parameters. 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           add53=T,add35=T) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-7-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Configuring the axis (part-2) 

 
 
We can also configure the resolution of the axis (albeit rather bluntly) using the **littleTicks** parameter. 
 
This will add additional axis tick marks between those shown by default. 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           littleTicks = TRUE) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-9-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Configuring the axis (part-3) 

 
 
By default the plot labels for the genome axis track are alternating below and above the line. 
 
We can further configure the axis labels using the **labelPos** parameter. 
 
Here we set the labelPos to be always below the axis 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           labelPos="below") 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-11-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Configuring the axis (part-4) 

 
 
In the previous plots we have produced a genomic axis which allows us to consider the position of the features within the linear genome. 
 
In some contexts we may be more interested in relative distances around and between the genomic features being displayed. 
 
We can configure the axis track to give us a relative representative of distance using the **scale** parameter. 
 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           scale=1,labelPos="below") 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-13-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Configuring the axis (part-4b) 

 
 
We may want to add only a part of the scale (such as with Google Maps) to allow the reviewer to get a sense of distance. 
 
We can specify how much of the total axis we wish to display as a scale using a value of 0 to 1 representing the proportion of scale to show. 
 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           scale=0.3) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-15-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Configuring the axis (part-4c) 

 
 
We can also provide numbers greater than 1 to the **scale** parameter which will determine, in absolute base pairs, the size of scale to display. 
 
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           scale=2500) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-17-1.png)&lt;!-- --&gt;
 
---
##Getting started with Gviz -- Axis and Regions of Interest (part-1) 

 
 
Previously we have seen how to highlight regions of interest in the scale bar for IGV. 
 
These "regions of interest" may be user defined locations which add context to the scale and the genomics data to be displayed (e.g. Domain boundaries such as topilogically associated domains) 
 
![ROI](../../imgs/igv_BookMarks.png) 
 
---
##Getting started with Gviz -- Axis and Regions of Interest (part-2) 

 
 
We can add "regions of interest" to the axis plotted by Gviz as we have done with IGV. 
 
To do this we will need to define some ranges to signify the positions of "regions of interest" in the linear context of our genome track. 
 
Since the plots have no apparent context for chromosomes (yet), we will use a IRanges object to specify "regions of interest" as opposed to the genome focused GRanges. 
 
You can see our material [here](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicIntervals_In_Bioconductor.html) on Bioconductor objects for more information on IRanges and GRanges. 
---
##Brief recap (Creating an IRanges) 

 
 
To create an IRanges object we will load the IRanges library and specify vectors of **start** and **end** parameters to the **IRanges** constructor function. 
 
 

```r
library(IRanges) 
regionsOfInterest &lt;- IRanges(start=c(140,5140),end=c(2540,7540)) 
names(regionsOfInterest) &lt;- c("ROI_1","ROI_2") 
regionsOfInterest 
```

```
## IRanges object with 2 ranges and 0 metadata columns:
##             start       end     width
##         &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   ROI_1       140      2540      2401
##   ROI_2      5140      7540      2401
```
---
##Getting started with Gviz -- Axis and Regions of Interest (part-3) 

 
 
Now we have our IRanges object representing our regions of interest we can include them in our axis. 
 
We will have to recreate our axis track to allow us to include these regions of interest. 
 
Once we have updated our GenomeAxisTrack object we can plot the axis with regions of interest included. 
 
 

```r
genomeAxis &lt;- GenomeAxisTrack(name="MyAxis", 
                              range = regionsOfInterest) 
plotTracks(genomeAxis,from=100,to=10100) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-20-1.png)&lt;!-- --&gt;
---
##Getting started with Gviz -- Axis and Regions of Interest (part-4) 

 
 
We include the names specified in the IRanges for the regions of interest within the axis plot by specify the **showID** parameter to TRUE. 
 
![](Viz_part1_files/figure-html/unnamed-chunk-21-1.png)&lt;!-- --&gt;
 

```r
plotTracks(genomeAxis,from=100,to=10100, 
           range=regionsOfInterest, 
           showId=T) 
```
 
---
##Plotting regions in Gviz - Data tracks 

Now we have some fine control of the axis, it follows that we may want some to display some actual data along side our axis and/or regions of interest. 
 
Gviz contains a general container for data tracks which can be created using the **DataTrack()** constructor function and associated object, **DataTrack**. 
 
Generally **DataTrack** may be used to display most data types with some work but best fits ranges with associated signal as a matrix (multiple regions) or vector (single sample). 
---
##Plotting regions in Gviz - Data tracks 

Lets update our IRanges object to have some score columns in the metadata columns. We can do this with the **mcols** function as shown in our Bioconductor material. 
 
 

```r
mcols(regionsOfInterest) &lt;- data.frame(Sample1=c(30,20),
                                       Sample2=c(20,200)) 
regionsOfInterest &lt;- GRanges(seqnames="chr5",
                             ranges = regionsOfInterest) 
regionsOfInterest 
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##         seqnames       ranges strand |   Sample1   Sample2
##            &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;
##   ROI_1     chr5 [ 140, 2540]      * |        30        20
##   ROI_2     chr5 [5140, 7540]      * |        20       200
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
 
---
##Plotting regions in Gviz - Data tracks 

 
 
Now we have the data we need, we can create a simple **DataTrack** object. 
 

```r
dataROI &lt;- DataTrack(regionsOfInterest) 
plotTracks(dataROI) 
```
 
![](Viz_part1_files/figure-html/unnamed-chunk-25-1.png)&lt;!-- --&gt;
---
##Plotting regions in Gviz - Data tracks 

 
 
As we have seen, **DataTrack** objects make use of IRanges/GRanges which are the central workhorse of Bioconductors HTS tools. 
 
This means we can take advantage of the many manipulations available in the Bioconductor tool set. 
 
Lets make use of rtracklayer's importing tools to retrieve coverage from a bigWig as a GRanges object 
 
 

```r
library(rtracklayer) 
allChromosomeCoverage &lt;- import.bw("../../Data/small_Sorted_SRR568129.bw",
                                   as="GRanges") 
class(allChromosomeCoverage)
```

```
## [1] "GRanges"
## attr(,"package")
## [1] "GenomicRanges"
```

---
##Plotting regions in Gviz - Data tracks 

Our GRanges from the bigWig contains intervals and their respective score (number of reads).


```r
allChromosomeCoverage 
```


```
## GRanges object with 248 ranges and 1 metadata column:
##         seqnames         ranges strand |     score
##            &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##     [1]     chr1 [1, 249250621]      * |         0
##     [2]     chr2 [1, 243199373]      * |         0
##     [3]     chr3 [1, 198022430]      * |         0
##     [4]     chr4 [1, 191154276]      * |         0
##     [5]     chr5 [1, 134887177]      * |         0
##     ...      ...            ...    ... .       ...
##   [244]    chr20 [1,  63025520]      * |         0
##   [245]    chr21 [1,  48129895]      * |         0
##   [246]    chr22 [1,  51304566]      * |         0
##   [247]     chrX [1, 155270560]      * |         0
##   [248]     chrY [1,  59373566]      * |         0
##   -------
##   seqinfo: 25 sequences from an unspecified genome
```

---
##Plotting regions in Gviz - Data tracks (part 4) 

 
Now we have our coverage as a GRanges object we can create our **DataTrack** object from this. 
 
Notice we specify the chromsome of interest in the **chromosome** parameter. 
 

```r
accDT &lt;- DataTrack(allChromosomeCoverage,chomosome="chr5") 
accDT 
```

```
## DataTrack 'DataTrack'
## | genome: NA
## | active chromosome: chrM
## | positions: 1
## | samples:1
## | strand: *
## There are 248 additional annotation features on 24 further chromosomes
##   chr1: 1
##   chr10: 1
##   chr11: 1
##   chr12: 1
##   chr13: 1
##   ...
##   chr7: 1
##   chr8: 1
##   chr9: 1
##   chrX: 1
##   chrY: 1
## Call seqlevels(obj) to list all available chromosomes or seqinfo(obj) for more detailed output
## Call chromosome(obj) &lt;- 'chrId' to change the active chromosome
```
 
---
##Plotting regions in Gviz - Data tracks (part 5) 

 
 
To plot data now using the plotTracks() function we will set the regions we wish to plot by specifying the chromsomes, start and end using the **chromosome**, **from** and **to** parameters. 
 
By default we will get a similar point plot to that seen before. 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-30-1.png)&lt;!-- --&gt;
 
---
##Plotting regions in Gviz - Data tracks (part 6) 

 
 
We can adjust the type of plots we want using the **type** argument. 
Here as with standard plotting we can specify **"l"** to get a line plot. 
 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5",type="l") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-31-1.png)&lt;!-- --&gt;
 
---
##Plotting regions in Gviz - Data tracks (part 6) 

 
Many other types of plots are available for the DataTracks. 
 
Including smoothed plots using "smooth". 
 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5",type="smooth") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-32-1.png)&lt;!-- --&gt;
 
 
---
##Plotting regions in Gviz - Data tracks (part 7) 

 
 
Histograms by specifying "h". 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5",type="h") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-33-1.png)&lt;!-- --&gt;
---
##Plotting regions in Gviz - Data tracks (part 8) 

 
Or filled/smoothed plots using "mountain". 
 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5",type="mountain") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-34-1.png)&lt;!-- --&gt;
 
---
##Plotting regions in Gviz - Data tracks (part 9) 

 
 
and even a Heatmap using "heatmap". 
 
Notice that Gviz will automatically produce the appropriate Heatmap scale. 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5",type="heatmap") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-35-1.png)&lt;!-- --&gt;
---
##Plotting regions in Gviz - Additional Parameters. 

 
 
As with all plotting functions in R, Gviz plots are highly customisable. 
 
Simple features such as point size and colour are easily set as for standard R plots using **cex** and **col** paramters. 
 

```r
plotTracks(accDT, 
           from=134887451,to=134888111, 
           chromosome="chr5", 
           col="red",cex=4) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-36-1.png)&lt;!-- --&gt;
 
---
##Putting track togethers - Axis and Data 

 
 
Now we have shown how to construct a data track and axis track we can put them together in one plot. 
 
To do this we simply provide the GenomeAxisTrack and DataTrack objects as vector the **plotTracks()** function. 
 
 

```r
plotTracks(c(accDT,genomeAxis), 
           from=134887451,to=134888111, 
           chromosome="chr5" 
           ) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-37-1.png)&lt;!-- --&gt;
---
##Putting track togethers - Ordering tracks in plot 

 
 
The order of tracks in the plot is simply defined by the order they are placed in the vector passed to **plotTracks()** 
 
 

```r
plotTracks(c(genomeAxis,accDT), 
           from=134887451,to=134888111, 
           chromosome="chr5" 
           ) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-38-1.png)&lt;!-- --&gt;
---
##Putting track togethers - Controling height of tracks in plot 

 
 
By default, Gviz will try and provide sensible track heights for your plots to best display your data. 
 
The track height can be controlled by providing a vector of relative heights to the **sizes** parameter of the **plotTracks()** function. 

---
##Putting track togethers - Controling height of tracks in plot 

If we want the axis to be 50% of the height of the Data track we specify the size for axis as 0.5 and that of data as 1. The order of sizes must match the order of objects they relate to. 
 
 

```r
plotTracks(c(genomeAxis,accDT), 
           from=134887451,to=134888111, 
           chromosome="chr5", 
           sizes=c(0.5,1) 
           ) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-39-1.png)&lt;!-- --&gt;
 

---
##Adding annotation to plots. 

Genomic annotation, such as Gene/Transcript models, play an important part of visualising genomics data in context. 

Gviz provides many routes for constructing genomic annotation using the **AnnotationTrack()** constructor function. 

```r
toGroup &lt;- GRanges(seqnames="chr5", 
        IRanges( 
          start=c(10,500,550,2000,2500), 
          end=c(300,800,850,2300,2800) 
        )) 
names(toGroup) &lt;- seq(1,5) 
toGroup 
```

```
## GRanges object with 5 ranges and 0 metadata columns:
##     seqnames       ranges strand
##        &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt;
##   1     chr5 [  10,  300]      *
##   2     chr5 [ 500,  800]      *
##   3     chr5 [ 550,  850]      *
##   4     chr5 [2000, 2300]      *
##   5     chr5 [2500, 2800]      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
##Adding annotation to plots. Grouping (part-1) 

 
 
Now we can create the **AnnotationTrack** object using the constructor. In contrast to the **DataTrack**, **AnnotationTrack** allows for the specification of feature groups. 

 
Here we provide a grouping to the **group** parameter in the **AnnotationTrack()** function. 
 

```r
annoT &lt;- AnnotationTrack(toGroup, 
                group = c("Ann1","Ann1", 
                          "Ann2", 
                          "Ann3","Ann3")) 
 
plotTracks(annoT) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-41-1.png)&lt;!-- --&gt;
 
---
##Adding annotation to plots. 

 
 
We can see the features are displayed grouped by lines. 
 
But if we want to see the names we must specify the group parameter by  using the **groupAnnotation** argument. 
 

```r
plotTracks(annoT,groupAnnotation = "group") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-42-1.png)&lt;!-- --&gt;
---
##Adding annotation to plots. Strands and direction. 

 
 
When we created the GRanges used here we did not specify any strand information. 
 

```r
strand(toGroup) 
```

```
## factor-Rle of length 5 with 1 run
##   Lengths: 5
##   Values : *
## Levels(3): + - *
```
 
When plotting annotation without strand a box is used to display features as seen in previous slides 
---
##Adding annotation to plots. Strands and direction (part-2). 

 
 
Now we can specify some strand information for the GRanges and replot. 
 
Arrows now indicate the strand which the features are on. 
 

```r
strand(toGroup) &lt;- c("+","+","*","-","-") 
annoT &lt;- AnnotationTrack(toGroup, 
                group = c("Ann1","Ann1", 
                          "Ann2", 
                          "Ann3","Ann3")) 
 
plotTracks(annoT, groupingAnnotation="group") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-44-1.png)&lt;!-- --&gt;
---
##Adding annotation to plots. Controlling the display density 

 
 
In the IGV course we saw how you could control the display density of certain tracks.  
 
Annotation tracks are often stored in files such as the general feature format (see our previous course).  
 
IGV allows us to control the density of these tracks in the view options by setting to "collapsed", "expanded" or "squished". 
 
Whereas "squished" and "expanded" maintains much of the information within the tracks, "collapsed" flattens overlapping features into a single displayed feature. 
 
 

---
##Adding annotation to plots. Controlling the display density (part 2) 

 
 
Here we have the same control over the display density of our annotation tracks. 
 
By default the tracks are stacked using the "squish" option to make best use of the available space. 
 

 

```r
plotTracks(annoT, groupingAnnotation="group",stacking="squish") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-47-1.png)&lt;!-- --&gt;
 
---
##Adding annotation to plots. Controlling the display density (part 3) 

 
 
By setting the **stacking** parameter to "dense", all overlapping features have been collapsed/flattened 
 

```r
plotTracks(annoT, groupingAnnotation="group",stacking="dense") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-48-1.png)&lt;!-- --&gt;
 
 
---
##Adding annotation to plots. Feature types. 

 
 
**AnnotationTrack** objects may also hold information on feature types. 
 
For gene models we may be use to feature types such as mRNA, rRNA, snoRNA etc. 
 
Here we can make use of feature types as well. 
 
We can set any feature types within our data using the **feature()** function. Here they are unset so displayed as unknown. 
 
 

```r
feature(annoT) 
```

```
## [1] "unknown" "unknown" "unknown" "unknown" "unknown" "unknown"
```
---
##Adding annotation to plots. Setting feature types. 

 
 
We can set our own feature types for the **AnnotationTrack** object using the same **feature()** function. 
 
We can choose any feature types we wish to define. 
 

```r
feature(annoT) &lt;- c(rep("Good",4),rep("Bad",2)) 
feature(annoT) 
```

```
## [1] "Good" "Good" "Good" "Good" "Bad"  "Bad"
```
---
##Adding annotation to plots. Display feature types. 

 
 
Now we have defined our feature types we can use this information within our plots. 
 
In GViz, we can directly specify attributes for individual feature types within our AnnotationTrack, in this example we add attributes for colour to be displayed. 
 
We specify the "Good" features as blue and the "Bad" features as red. 
 

```r
plotTracks(annoT, featureAnnotation = "feature", 
           groupAnnotation = "group", 
           Good="Blue",Bad="Red") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-51-1.png)&lt;!-- --&gt;
 
---
##GeneRegionTrack 

 
id:grtrack 
 
We have seen how we can display complex annotation using the **AnnotationTrack** objects. 
 
For gene models Gviz contains a more specialised object, the **GeneRegionTrack** object. 
 
The **GeneRegionTrack** object contains additional parameters and display options specific for the display of gene models. 
 
Lets start by looking at the small gene model set stored in the **Gviz** package. 
 
 

```r
data(geneModels) 
head(geneModels) 
```

```
##   chromosome    start      end width strand feature            gene
## 1       chr7 26591441 26591829   389      + lincRNA ENSG00000233760
## 2       chr7 26591458 26591829   372      + lincRNA ENSG00000233760
## 3       chr7 26591515 26591829   315      + lincRNA ENSG00000233760
## 4       chr7 26594428 26594538   111      + lincRNA ENSG00000233760
## 5       chr7 26594428 26596819  2392      + lincRNA ENSG00000233760
## 6       chr7 26594641 26594733    93      + lincRNA ENSG00000233760
##              exon      transcript     symbol
## 1 ENSE00001693369 ENST00000420912 AC004947.2
## 2 ENSE00001596777 ENST00000457000 AC004947.2
## 3 ENSE00001601658 ENST00000430426 AC004947.2
## 4 ENSE00001792454 ENST00000457000 AC004947.2
## 5 ENSE00001618328 ENST00000420912 AC004947.2
## 6 ENSE00001716169 ENST00000457000 AC004947.2
```
---
##GeneRegionTrack 

 
 

```
##   chromosome    start      end width strand feature            gene
## 1       chr7 26591441 26591829   389      + lincRNA ENSG00000233760
## 2       chr7 26591458 26591829   372      + lincRNA ENSG00000233760
## 3       chr7 26591515 26591829   315      + lincRNA ENSG00000233760
## 4       chr7 26594428 26594538   111      + lincRNA ENSG00000233760
## 5       chr7 26594428 26596819  2392      + lincRNA ENSG00000233760
## 6       chr7 26594641 26594733    93      + lincRNA ENSG00000233760
##              exon      transcript     symbol
## 1 ENSE00001693369 ENST00000420912 AC004947.2
## 2 ENSE00001596777 ENST00000457000 AC004947.2
## 3 ENSE00001601658 ENST00000430426 AC004947.2
## 4 ENSE00001792454 ENST00000457000 AC004947.2
## 5 ENSE00001618328 ENST00000420912 AC004947.2
## 6 ENSE00001716169 ENST00000457000 AC004947.2
```
 
We can see that this data.frame contains information on start, end , chromosome and strand of feature needed to position features in a linear genome. 
 
Also included are a feature type column named "feature" and columns containing additional metadata to group by - "gene","exon","transcript","symbol". 
 
---
##GeneRegionTrack - Setting up the gene model track. 

 
 
We can define a GeneRegionTrack as we would all other tracktypes. Here we provide a genome name, chromosome of interest and a name for the track. 
 
 

```r
grtrack &lt;- GeneRegionTrack(geneModels, genome = "hg19", 
                           chromosome = "chr7", 
                           name = "smallRegions") 
plotTracks(grtrack) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-54-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Setting up the gene model track. 

 
 

```r
plotTracks(grtrack) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-55-1.png)&lt;!-- --&gt;
 
We can see that features here are rendered slightly differently to those in an **AnnotationTrack** object. 
 
Here direction is illustrated by arrows in introns and unstranslated regions are shown as narrower boxes. 
 
---
##GeneRegionTrack - Specialised labelling. 

 
 
Since gene models typically contain exon, transcript and gene level annotation we can specify how we wish to annotate our plots by using the **transcriptAnnotation** and **exonAnnotation** parameters. 
 
To label all transcripts by the gene annotation we specify the gene column to the **transcriptAnnotation** parameter. 
 
 

```r
plotTracks(grtrack,transcriptAnnotation="gene") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-56-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialised labelling. 

 
 
Similarly we can label transcripts by their individual transcript names. 
 

```r
plotTracks(grtrack,transcriptAnnotation="transcript") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-57-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialised labelling. 

 
 
Or we can label using the **transcriptAnnotation** object by any arbitary column where there is one level per transcript. 
 

```r
plotTracks(grtrack,transcriptAnnotation="symbol") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-58-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialised labelling of exons. 

 
 
As with transcripts we can label individual features using the **exonAnnotation** parameter by any arbitary column where there is one level per feature/exon. 
 

```r
plotTracks(grtrack,exonAnnotation="exon",from=26677490,to=26686889,cex=0.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-59-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialized display density for gene models. 

 
 
We saw that we can control the display density when plotting **AnnotationTrack** objects. 
 
We can control the display density of GeneRegionTracks in the same manner. 
 

```r
plotTracks(grtrack, stacking="dense") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-60-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialized display density for gene models. 

 
 
However, since the **GeneRegionTrack** object is a special class of the **AnnotationTrack** object we have special parameter for dealing with display density of transcripts. 
 
The **collapseTranscripts** parameter allows us a finer degree of control than that seen with **stacking** parameter. 
 
Here we set **collapseTranscripts** to be true inorder to merge all overlapping transcripts.  
 

```r
plotTracks(grtrack, collapseTranscripts=T, 
           transcriptAnnotation = "symbol") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-61-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialized display density for gene models. 

 
 
Collapsing using the **collapseTranscripts** has summarised our transcripts into their respective gene boundaries. 
 
We have however lost information on the strand of transcripts. To retain this information we need to specify a new shape for our plots using the **shape** parameter. To capture direction we use the "arrow" shape 
 

```r
plotTracks(grtrack, collapseTranscripts=T, 
           transcriptAnnotation = "symbol", 
           shape="arrow") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-62-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Specialized display density for gene models. 

 
 
The **collapseTranscripts** function also allows us some additional options by which to collapse our transcripts. 
 
These methods maintain the intron information in the gene model and so get us closer to reproducing the "collapsed" feature in IGV. 
 
Here we may collapse transcripts to the "longest". 
 

```r
plotTracks(grtrack, collapseTranscripts="longest", 
           transcriptAnnotation = "symbol") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-63-1.png)&lt;!-- --&gt;
 
---
##GeneRegionTrack - Specialized display density for gene models. 

 
 
Or we may specify to **collapseTranscripts** function to collapse by "meta". 
 
The "meta" option shows us a composite, lossless illustration of the gene models closest to that seen in "collapsed" IGV tracks. 
 
Here importantly all exon information is retained. 
 

```r
plotTracks(grtrack, collapseTranscripts="meta", 
           transcriptAnnotation = "symbol") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-64-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Building your own gene models. 

 
 
We have seen in previous material how gene models are organised in Bioconductor using the **TxDB** objects. 
 
Gviz may be used in junction with **TxDB** objects to construct the **GeneRegionTrack** objects.  
 
We saw in the Bioconductor and ChIPseq course that many genomes have pre-build gene annotation within the respective TxDB libraries. Here we will load a **TxDb** for hg19 from the  **TxDb.Hsapiens.UCSC.hg19.knownGene** library. 

```r
library(TxDb.Hsapiens.UCSC.hg19.knownGene) 
 
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene 
txdb 
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
```
---
##GeneRegionTrack - Building your own gene models from a TxDb. 

 
 
Now we have loaded our **TxDb** object and assigned it to *txdb*. We can use this **TxDb** object to construct our **GeneRegionTrack**. Here we focus on chromosome 7 again. 
 

```r
customFromTxDb &lt;- GeneRegionTrack(txdb,chromosome="chr7") 
head(customFromTxDb) 
```

```
## GeneRegionTrack 'GeneRegionTrack'
## | genome: hg19
## | active chromosome: chr7
## | annotation features: 6
```
---
##GeneRegionTrack - Building your own gene models from a TxDb. 

 
 
With our new **GeneRegionTrack** we can now reproduce the gene models using the Bioconductor TxDb annotation. 
 
Here the annotation is different but transcripts overlapping uc003syc are our SKAP2 gene. 
 

```r
plotTracks(customFromTxDb, 
           from=26591341,to=27034958, 
           transcriptAnnotation="gene") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-67-1.png)&lt;!-- --&gt;
---
##GeneRegionTrack - Building your own gene models from a GFF. 

 
 
Now by combining the ability to create our own **TxDb** objects from GFFs we can create a very custom GeneRegionTrack from a GFF file. Here i use the [UCSC table browser](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=676227945_cx1yoPC6BgjmGcjZrorhNkXul8xS&amp;clade=mammal&amp;org=&amp;db=hg19&amp;hgta_group=genes&amp;hgta_track=knownGene&amp;hgta_table=knownGene&amp;hgta_regionType=genome&amp;position=&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=) to extract a GTF of interest.
 
 

```r
library(GenomicFeatures) 
ensembleGTF &lt;- "~/Downloads/hg19.gtf.gz"
txdbFromGFF &lt;- makeTxDbFromGFF(file = ensembleGTF) 
```

```
## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
```

```r
customFromTxDb &lt;- GeneRegionTrack(txdbFromGFF,chromosome="chr7") 
plotTracks(customFromTxDb, 
           from=26591341,to=27034958, 
           transcriptAnnotation="gene") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-68-1.png)&lt;!-- --&gt;
---
##Exercises 

 
Time for exercises! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/exercises/AnnotationAndGeneRegionTrack_Exercies.html) 
 
---
##Solutions 

 
Time for solutions! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/solutions/AnnotationAndGeneRegionTrack_Solutions.html) 
 
 
---
##SequenceTracks 

 
id:seqtrack 
 
 
When displaying genomics data it can be important to illustrate the underlying sequence for the genome being viewed. 
 
Gviz uses **SequenceTrack** objects to handle displaying sequencing information. 
 
First we need to get some  sequence information for our genome of interest to display. Here we will use one of the **BSgenome** packages specific for hg19 - **BSgenome.Hsapiens.UCSC.hg19**. This contains the full sequence for hg19 as found in UCSC 
 

```r
library(BSgenome.Hsapiens.UCSC.hg19) 
```

```
## Loading required package: BSgenome
```

```r
BSgenome.Hsapiens.UCSC.hg19[["chr7"]] 
```

```
##   159138663-letter "DNAString" instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
```
---
##SequenceTracks - From a BSgenome object 

 
 
We can create a **SequenceTrack** object straight from this **BSgenome** object using the **SequenceTrack()** constructor.  
 
We can then plot this **SequenceTrack**, as with all tracks, using the **plotTracks()** functions. Here we specify the *from*, *to* and *chromosome* parameters to select a region to display. 
 
 

```r
sTrack &lt;- SequenceTrack(Hsapiens) 
plotTracks(sTrack,from=134887024,to=134887074, 
           chromosome = "chr7",cex=2.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-70-1.png)&lt;!-- --&gt;
---
##SequenceTracks - From a DNAstringset object 

 
 
We can also specify a DNAstringset object which we have encountered in the [Bioconductor](https://mrccsc.github.io/Bioconductor/) and [ChIP-seq](https://mrccsc.github.io/ChIPseq_short/) courses. 
 

```r
dsSet &lt;- DNAStringSet(Hsapiens[["chr7"]]) 
names(dsSet) &lt;- "chr7" 
sTrack &lt;- SequenceTrack(dsSet) 
plotTracks(sTrack,from=134887024,to=134887074, 
           chromosome = "chr7",cex=2.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-71-1.png)&lt;!-- --&gt;
 
---
##SequenceTracks - From a DNAstringset object 

 
 
We can also create our custom SequenceTrack from a [Fasta](https://mrccsc.github.io/genomicFormats) file. 
 
Here we use an example containing only the sequence around the region we are looking at to save space. Since the sequence is only of the region of interest we need specify the sequence limits for the *from* and *to* arguments. With completer fasta files, **from** and **to** would be set as for other **SequenceTrack** examples. 
 

 
 

```r
sTrack &lt;- SequenceTrack("../../Data/chr7Short.fa") 
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7",cex=3) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-73-1.png)&lt;!-- --&gt;
---
##SequenceTracks - Displaying complement sequence 

 
 
As with IGV, the sequence can be displayed as its complement. This is performed here by setting the **complement** argument to the **plotTracks()** function to TRUE/T. 
 

```r
sTrack &lt;- SequenceTrack("../../Data/chr7Short.fa") 
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7",complement=T,cex=3) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-74-1.png)&lt;!-- --&gt;
---
##SequenceTracks - Displaying strand information 

 
 
We can also add 5' to 3' direction as we have for plotting **GenomeAxisTrack**  objects using the **add53** parameter. This allows for a method to illustrate the strand of the sequence being diplayed. 
 

```r
sTrack &lt;- SequenceTrack("../../Data/chr7Short.fa") 
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7",complement=F, 
           add53=T,cex=2.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-75-1.png)&lt;!-- --&gt;
---
##SequenceTracks - Displaying strand information 

 
 
Notice the 5' and 3' labels have swapped automatically when we have specified the complement sequence. 
 

```r
sTrack &lt;- SequenceTrack("../../Data/chr7Short.fa") 
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7",complement=T, 
           add53=T,cex=2.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-76-1.png)&lt;!-- --&gt;
---
##SequenceTracks - Controlling base display size 

 
 
We can control the size of bases with the **cex** parameter, as with the standard R plotting.  
 
An interesting feature of this is that when plotted bases overlap, Gviz will provide a colour representation of bases instead of the bases' characters. 
 
 

```r
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7",cex=2.5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-77-1.png)&lt;!-- --&gt;

```r
plotTracks(sTrack,from=1,to=50, 
           chromosome = "chr7", 
           cex=5) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-77-2.png)&lt;!-- --&gt;
 
---
##AlignmentsTrack.  

 
id:alignments 
 
So far we have displayed summarised genomics data using GRange objects or GRanges with associated metadata. 
 
A prominent feature of Gviz is that it can work with genomic alignments, providing methods to generate graphical summaries on the fly. 
 
Genomic alignments are stored in Gviz within the AlignmentsTrack object. 
 
Here we can read genomic alignments in from a BAM file, see our file formats course material, by specifying its location. 
 
 

```r
   peakReads &lt;- AlignmentsTrack("../../Data/small_Sorted_SRR568129.bam") 
   peakReads 
```

```
## ReferenceAlignmentsTrack 'AlignmentsTrack'
## | genome: NA
## | active chromosome: chrNA
## | referenced file: ../../Data/small_Sorted_SRR568129.bam
## | mapping: id=id, cigar=cigar, mapq=mapq, flag=flag, isize=isize, groupid=groupid, status=status, md=md, seq=seq
```
---
##AlignmentsTrack.  Plotting Aligned Reads in Gviz 

 
 
The **AlignmentsTrack** object can be plotted in the same manner as tracks using **plotTracks()** function. 
 
Since the BAM file may contain information from all chromosomes we need to specify a chromsome to plot in the **chromosome** parameter and here we specify the **from** and **to** parameters too. 
 

```r
   plotTracks(peakReads, 
              chromosome="chr5", 
              from=135312577, 
              to=135314146) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-79-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Plotting Aligned Reads in Gviz 

 
 

```r
   plotTracks(peakReads, 
              chromosome="chr5", 
              from=135312577, 
              to=135314146) 
```

![](Viz_part1_files/figure-html/unnamed-chunk-80-1.png)&lt;!-- --&gt;
 
 
By default **AlignmentTrack**s are rendered as the both the reads themselves and the calculated coverage/signal depth from these reads. 
 
Reads, as with AnnotationTrack objects, show the strand of the aligned read by the direction of the arrow. 
---
##AlignmentsTrack.  Plotting Aligned Reads in Gviz 

 
 
The type of plot/plots produced can be controlled by the **type** argument as we have done for **DataTrack** objects. 
 
The valid types of plots for AlignmentsTrack objects are "pileup", "coverage" and "sashimi" *(We've come across sashimi plots before)*.  
 
The type "pileup" displays just the reads. 
 

```r
   plotTracks(peakReads, 
              chromosome="chr5", 
              from=135312577, 
              to=135314146, 
              type="pileup") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-81-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Plotting Aligned Reads in Gviz 

 
 
The type "coverage" displays just the coverage (depth of signal over genomic positions) calculated from the genomic alignments. 
 

```r
   plotTracks(peakReads, 
              chromosome="chr5", 
              from=135312577, 
              to=135314146, 
              type="coverage") 
```

![](Viz_part1_files/figure-html/unnamed-chunk-82-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Plotting Aligned Reads in Gviz 

 
 
As we have seen the default display is a combination of "pileup" and "coverage". 
 
We can provide multiple *type* arguments to the **plotTracks()** function as a vector of valid types. The order in vector *here* does not affect the display order in panels. 
 
![](Viz_part1_files/figure-html/unnamed-chunk-83-1.png)&lt;!-- --&gt;
 
 
---
##AlignmentsTrack.  Sashimi plots 

 
 
We have seen [sashimi plots in IGV](http://mrccsc.github.io/IGV_course/igv.html#/53) when reviewing RNA-seq data. 
 
Sashimi plots display the strength of signal coming from reads spanning splice junctions and so can act to illustrate changes in exon usage between samples. 
 
In IGV, we previous made use of the **BodyMap** data to show alternative splicing of an exon between heart and liver. 
 
&lt;div align="center"&gt; 
&lt;img src="../../imgs/IGV_SplicingExample.png" alt="offset" height="600" width="1000"&gt; 
&lt;/div&gt; 
---
##AlignmentsTrack.  Sashimi plots in Gviz 

 
 
To recapitulate this plot, we have retrieved the subsection of **BodyMap** data as BAM files. 
 
First we must create two **AlignmentsTrack** objects, one for each tissue's BAM file of aligned reads.  
 
In this case since we are working with paired-end reads we must specify this by setting the **isPaired** parameter to TRUE 
 
 

```r
heartReads &lt;- AlignmentsTrack("../../Data/heart.bodyMap.bam", 
                           isPaired = TRUE) 
liverReads &lt;- AlignmentsTrack("../../Data/liver.bodyMap.bam",  
                           isPaired = TRUE) 
 
liverReads 
```

```
## ReferenceAlignmentsTrack 'AlignmentsTrack'
## | genome: NA
## | active chromosome: chrNA
## | referenced file: ../../Data/liver.bodyMap.bam
## | mapping: id=id, cigar=cigar, mapq=mapq, flag=flag, isize=isize, groupid=groupid, status=status, md=md, seq=seq
```
---
##AlignmentsTrack.  Sashimi plots in Gviz 

 
 
As with **DataTrack** objects we can combine the **AlignmentTrack**s as a vector for plotting with the **plotTracks()** function. 
 
By default we will display the reads and calculated coverage. Here the paired reads and split reads are illustrated by thick and thin lines respectively. 
 

```r
plotTracks(c(heartReads,liverReads), 
           chromosome="chr12", 
           from=98986825,to=98997877) 
```

![](Viz_part1_files/figure-html/bb-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Sashimi plots in Gviz 

 
 
To reproduce a plot similar to that in IGV we can simply include the "sashimi" type in the **type** parameter vector, here alongside "coverage"  
 

```r
plotTracks(c(heartReads,liverReads), 
           chromosome="chr12", 
           from=98986825, 
           to=98997877, 
           type=c("coverage","sashimi")) 
```

![](Viz_part1_files/figure-html/cc-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Highlighting genomic alignment information. 

 
 
The **AlignmentTrack** object allows for specific parameters controlling how reads are displayed to be passed to the **plotTracks()** function. 
 
A few useful parameters are **col.gaps** and **col.mates** or **lty.gap** and **lty.mates** which will allow us to better distinguish between gapped alignments (split reads) and gaps between read pairs respectively. 
 
 

```r
plotTracks(c(liverReads), 
           chromosome="chr12", 
           from=98986825,to=98997877, 
           col.gap="Red",col.mate="Blue") 
```

![](Viz_part1_files/figure-html/dd-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Highlighting genomic alignment information. 

 
 
Similarly using lty.gap and lty.mate parameters.  
 

```r
plotTracks(c(liverReads), 
           chromosome="chr12", 
           from=98986825,to=98997877, 
           lty.gap=2,lty.mate=1) 
```

![](Viz_part1_files/figure-html/ee-1.png)&lt;!-- --&gt;
 
Line width may also be controlled with lwd.gap and lwd.mate parameters continuing the similarities to Base R plotting. 
---
##AlignmentsTrack.  Highlighting mismatches to reference. 

 
 
A common purpose in visualising alignment data in broswers is review information relating to mismatches to the genome which may be related to SNPs. 
 
In order to highlight mismatches to the genome reference sequence we must first provide some information on the reference sequence. 
 
One method for this is to attach sequence information to the **AlignmentsTrack** itself by providing a **SequenceTrack** object to **referenceSequence** parameter in the **AlignmentsTrack()** constructor. Here we can use the **SequenceTrack** object we made earlier. 
 

```r
sTrack &lt;- SequenceTrack(Hsapiens) 
heartReads &lt;- AlignmentsTrack("../../Data/heart.bodyMap.bam", 
                           isPaired = TRUE, 
                           referenceSequence=sTrack) 
```
---
##AlignmentsTrack.  Highlighting mismatches to reference. 

 
 
Now when we can replot the pileup of reads where mismatches in the reads are highlighted. 
 

```r
plotTracks(heartReads, 
           chromosome="chr12", 
           from=98987800,to=98987977, 
           type="pileup") 
```

![](Viz_part1_files/figure-html/gg-1.png)&lt;!-- --&gt;
---
##AlignmentsTrack.  Highlighting mismatches to reference. 

 
 
We could also specify the SequenceTrack in the **plotTracks()** function as shown for the liver reads example here. Here we simply include the relevant **SequenceTrack** object as a track to be plotted  alongside the BAM. 
 

```r
plotTracks(c(liverReads,sTrack), 
           chromosome="chr12", 
           from=98987800, 
           to=98987977, 
           type="pileup") 
```

![](Viz_part1_files/figure-html/hh-1.png)&lt;!-- --&gt;
 
---
##Exercises 

 
 
 
Time for exercises! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/exercises/AlignmentTrack_Exercises.html) 
 
---
##Solutions 

 
 
 
Time for solutions! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/solutions/AlignmentTrack_Solutions.html) 
 
 
 
---
##Bringing in External data. 

 
id: externaldata 
 
**Gviz** has functions to allow us to import data from external repositories and databases. 
 
As in the IGV course, visualising genomics data in the context of additional genome information and external data held at these repositories provides a deeper insight into our own data. 
 
In this course we will look at two main methods of querying external databases- 
 
* The **BiomartGeneRegionTrack** object and constructor. 
* The **UcscTrack** object and constructor 
 
---
##Bringing in External data. Gene models through Biomart 

 
 
We have previously seen how we can use the **biomaRt** Bioconductor package to programatically query various Biomarts [(see our previous material)](https://mrccsc.github.io/Bioconductor/). 
 
**Gviz** allows us to both query Biomarts and automatically create a GeneRegionTrack using the **BiomartGeneRegionTrack** objects and **BiomartGeneRegionTrack()** constructor. 
---
##Bringing in External data. Gene models through Biomart 

 
 
Here we construct a simple **BiomartGeneRegionTrack** object using the parameters to define locations of interest - "chromsome", "start","end","genome" as well as the Biomart to use, in this case Ensembl by setting the **name** parameter. 
 

```r
bgrTrack &lt;- BiomartGeneRegionTrack(genome="hg19", 
                                   start=26591341, 
                                   end=27034958, 
                                   chromosome = "chr7", 
                                   name="ENSEMBL") 
```
---
##Bringing in External data. Gene models through Biomart 

 
 
We can then plot the BiomartGeneRegionTrack as we have previous GeneRegionTracks. 
 

```r
plotTracks(bgrTrack) 
```

![](Viz_part1_files/figure-html/b-1.png)&lt;!-- --&gt;
---
##Bringing in External data. Gene models through Biomart 

 
 
We can also specify filters in the **BiomartGeneRegionTrack()** constructor using the **filter** parameter. 
 
**Gviz** uses the **BiomaRt** Bioconductor package to query the Biomarts so we can apply the same filters as in **BiomaRt** (which we saw in our earlier material). 
 

```r
library(biomaRt) 
mart = useMart("ensembl", dataset="hsapiens_gene_ensembl") 
listFilters(mart) 
```

```
##                                      name
## 1                         chromosome_name
## 2                                   start
## 3                                     end
## 4                              band_start
## 5                                band_end
## 6                            marker_start
## 7                              marker_end
## 8                           encode_region
## 9                                  strand
## 10                     chromosomal_region
## 11                              with_ccds
## 12                            with_chembl
## 13          with_clone_based_ensembl_gene
## 14    with_clone_based_ensembl_transcript
## 15                            with_dbass3
## 16                            with_dbass5
## 17                 with_ens_hs_transcript
## 18                with_ens_hs_translation
## 19             with_entrezgene_trans_name
## 20                              with_embl
## 21                      with_arrayexpress
## 22                            with_genedb
## 23                                with_go
## 24                        with_goslim_goa
## 25                              with_ottp
## 26                              with_hgnc
## 27                   with_hgnc_trans_name
## 28                               with_hpa
## 29                        with_protein_id
## 30                       with_kegg_enzyme
## 31                      with_ens_lrg_gene
## 32                with_ens_lrg_transcript
## 33                            with_merops
## 34                          with_mim_gene
## 35                        with_mim_morbid
## 36                           with_mirbase
## 37                with_mirbase_trans_name
## 38                        with_entrezgene
## 39                               with_pdb
## 40                          with_reactome
## 41                     with_reactome_gene
## 42               with_reactome_transcript
## 43                       with_refseq_mrna
## 44             with_refseq_mrna_predicted
## 45                      with_refseq_ncrna
## 46            with_refseq_ncrna_predicted
## 47                    with_refseq_peptide
## 48          with_refseq_peptide_predicted
## 49                              with_rfam
## 50                   with_rfam_trans_name
## 51                        with_rnacentral
## 52                              with_ucsc
## 53                           with_uniparc
## 54                        with_uniprot_gn
## 55                  with_uniprotswissprot
## 56                   with_uniprotsptrembl
## 57                  with_vega_translation
## 58                          with_wikigene
## 59                        ensembl_gene_id
## 60                ensembl_gene_id_version
## 61                  ensembl_transcript_id
## 62          ensembl_transcript_id_version
## 63                     ensembl_peptide_id
## 64             ensembl_peptide_id_version
## 65                        ensembl_exon_id
## 66                     external_gene_name
## 67               external_transcript_name
## 68                                   ccds
## 69                                 chembl
## 70               clone_based_ensembl_gene
## 71         clone_based_ensembl_transcript
## 72                            dbass3_name
## 73                              dbass3_id
## 74                            dbass5_name
## 75                              dbass5_id
## 76                      ens_hs_transcript
## 77                     ens_hs_translation
## 78                  entrezgene_trans_name
## 79                                   embl
## 80                           arrayexpress
## 81                                 genedb
## 82                                     go
## 83                             goslim_goa
## 84                                   ottp
## 85                                hgnc_id
## 86                            hgnc_symbol
## 87                        hgnc_trans_name
## 88                                    hpa
## 89                             protein_id
## 90                            kegg_enzyme
## 91                           ens_lrg_gene
## 92                     ens_lrg_transcript
## 93                                 merops
## 94                     mim_gene_accession
## 95                   mim_morbid_accession
## 96                      mirbase_accession
## 97                             mirbase_id
## 98                     mirbase_trans_name
## 99                             entrezgene
## 100                                   pdb
## 101                              reactome
## 102                         reactome_gene
## 103                   reactome_transcript
## 104                           refseq_mrna
## 105                 refseq_mrna_predicted
## 106                          refseq_ncrna
## 107                refseq_ncrna_predicted
## 108                        refseq_peptide
## 109              refseq_peptide_predicted
## 110                                  rfam
## 111                       rfam_trans_name
## 112                            rnacentral
## 113                                  ucsc
## 114                               uniparc
## 115                            uniprot_gn
## 116                      uniprotswissprot
## 117                       uniprotsptrembl
## 118                      vega_translation
## 119                         wikigene_name
## 120                           wikigene_id
## 121                     with_affy_hc_g110
## 122                    with_affy_hg_focus
## 123                    with_affy_hg_u133a
## 124                  with_affy_hg_u133a_2
## 125                    with_affy_hg_u133b
## 126              with_affy_hg_u133_plus_2
## 127                     with_affy_hg_u95a
## 128                   with_affy_hg_u95av2
## 129                     with_affy_hg_u95b
## 130                     with_affy_hg_u95c
## 131                     with_affy_hg_u95d
## 132                     with_affy_hg_u95e
## 133                     with_affy_hta_2_0
## 134              with_affy_huex_1_0_st_v2
## 135                    with_affy_hugenefl
## 136            with_affy_hugene_1_0_st_v1
## 137            with_affy_hugene_2_0_st_v1
## 138                   with_affy_primeview
## 139                    with_affy_u133_x3p
## 140                  with_agilent_cgh_44b
## 141                  with_agilent_gpl6848
## 142    with_agilent_sureprint_g3_ge_8x60k
## 143 with_agilent_sureprint_g3_ge_8x60k_v2
## 144              with_agilent_wholegenome
## 145     with_agilent_wholegenome_4x44k_v1
## 146     with_agilent_wholegenome_4x44k_v2
## 147                with_codelink_codelink
## 148           with_illumina_humanht_12_v3
## 149           with_illumina_humanht_12_v4
## 150           with_illumina_humanref_8_v3
## 151            with_illumina_humanwg_6_v1
## 152            with_illumina_humanwg_6_v2
## 153            with_illumina_humanwg_6_v3
## 154                 with_phalanx_onearray
## 155                          affy_hc_g110
## 156                         affy_hg_focus
## 157                         affy_hg_u133a
## 158                       affy_hg_u133a_2
## 159                         affy_hg_u133b
## 160                   affy_hg_u133_plus_2
## 161                          affy_hg_u95a
## 162                        affy_hg_u95av2
## 163                          affy_hg_u95b
## 164                          affy_hg_u95c
## 165                          affy_hg_u95d
## 166                          affy_hg_u95e
## 167                          affy_hta_2_0
## 168                   affy_huex_1_0_st_v2
## 169                         affy_hugenefl
## 170                 affy_hugene_1_0_st_v1
## 171                 affy_hugene_2_0_st_v1
## 172                        affy_primeview
## 173                         affy_u133_x3p
## 174                       agilent_cgh_44b
## 175                       agilent_gpl6848
## 176         agilent_sureprint_g3_ge_8x60k
## 177      agilent_sureprint_g3_ge_8x60k_v2
## 178                   agilent_wholegenome
## 179          agilent_wholegenome_4x44k_v1
## 180          agilent_wholegenome_4x44k_v2
## 181                     codelink_codelink
## 182                illumina_humanht_12_v3
## 183                illumina_humanht_12_v4
## 184                illumina_humanref_8_v3
## 185                 illumina_humanwg_6_v1
## 186                 illumina_humanwg_6_v2
## 187                 illumina_humanwg_6_v3
## 188                      phalanx_onearray
## 189         transcript_count_greater_than
## 190            transcript_count_less_than
## 191                               biotype
## 192                    transcript_biotype
## 193                                source
## 194                     transcript_source
## 195                        transcript_tsl
## 196              transcript_gencode_basic
## 197                     transcript_appris
## 198                 phenotype_description
## 199                      phenotype_source
## 200                        go_parent_term
## 201                        go_parent_name
## 202                      go_evidence_code
## 203                 with_hsapiens_paralog
## 204                 with_mspretus_homolog
## 205                   with_vpacos_homolog
## 206                 with_pformosa_homolog
## 207               with_cpalliatus_homolog
## 208            with_acarolinensis_homolog
## 209            with_dnovemcinctus_homolog
## 210                   with_rbieti_homolog
## 211             with_sboliviensis_homolog
## 212                with_ppaniscus_homolog
## 213                  with_caperea_homolog
## 214               with_ogarnettii_homolog
## 215            with_cintestinalis_homolog
## 216                with_csavignyi_homolog
## 217                 with_celegans_homolog
## 218               with_ccapucinus_homolog
## 219                   with_fcatus_homolog
## 220               with_amexicanus_homolog
## 221                  with_ggallus_homolog
## 222             with_ptroglodytes_homolog
## 223               with_cchok1gshd_homolog
## 224                  with_ccrigri_homolog
## 225                with_psinensis_homolog
## 226                  with_gmorhua_homolog
## 227               with_lchalumnae_homolog
## 228               with_pcoquereli_homolog
## 229                  with_btaurus_homolog
## 230            with_mfascicularis_homolog
## 231              with_fdamarensis_homolog
## 232                   with_odegus_homolog
## 233              with_cfamiliaris_homolog
## 234               with_ttruncatus_homolog
## 235             with_mleucophaeus_homolog
## 236           with_aplatyrhynchos_homolog
## 237                with_lafricana_homolog
## 238                    with_mfuro_homolog
## 239              with_falbicollis_homolog
## 240            with_dmelanogaster_homolog
## 241                with_trubripes_homolog
## 242              with_nleucogenys_homolog
## 243                  with_chircus_homolog
## 244                 with_mauratus_homolog
## 245               with_rroxellana_homolog
## 246                 with_ggorilla_homolog
## 247               with_cporcellus_homolog
## 248               with_eeuropaeus_homolog
## 249                with_ecaballus_homolog
## 250                with_pcapensis_homolog
## 251                   with_dordii_homolog
## 252                 with_pmarinus_homolog
## 253                 with_jjaculus_homolog
## 254                with_etelfairi_homolog
## 255                with_clanigera_homolog
## 256               with_anancymaae_homolog
## 257                 with_mmulatta_homolog
## 258                 with_cjacchus_homolog
## 259                 with_olatipes_homolog
## 260                with_pvampyrus_homolog
## 261               with_mlucifugus_homolog
## 262                with_mmusculus_homolog
## 263                 with_mmurinus_homolog
## 264                  with_hfemale_homolog
## 265                    with_hmale_homolog
## 266                 with_pbairdii_homolog
## 267                  with_panubis_homolog
## 268               with_mdomestica_homolog
## 269                  with_pabelii_homolog
## 270             with_amelanoleuca_homolog
## 271                  with_sscrofa_homolog
## 272              with_mnemestrina_homolog
## 273                with_oprinceps_homolog
## 274               with_xmaculatus_homolog
## 275                with_oanatinus_homolog
## 276             with_mochrogaster_homolog
## 277               with_ocuniculus_homolog
## 278              with_rnorvegicus_homolog
## 279                  with_mcaroli_homolog
## 280              with_scerevisiae_homolog
## 281                   with_oaries_homolog
## 282                 with_saraneus_homolog
## 283                  with_mpahari_homolog
## 284               with_choffmanni_homolog
## 285                    with_catys_homolog
## 286                with_loculatus_homolog
## 287        with_itridecemlineatus_homolog
## 288               with_gaculeatus_homolog
## 289                with_csyrichta_homolog
## 290                with_sharrisii_homolog
## 291            with_tnigroviridis_homolog
## 292               with_oniloticus_homolog
## 293               with_tbelangeri_homolog
## 294               with_mgallopavo_homolog
## 295                  with_ngalili_homolog
## 296                 with_csabaeus_homolog
## 297                 with_neugenii_homolog
## 298              with_xtropicalis_homolog
## 299                 with_tguttata_homolog
## 300                   with_drerio_homolog
## 301                         with_interpro
## 302                           with_family
## 303                     with_sifts_import
## 304                              with_cdd
## 305                           with_gene3d
## 306                            with_hamap
## 307                       with_mobidblite
## 308                           with_ncoils
## 309                       with_hmmpanther
## 310                             with_pfam
## 311                            with_pirsf
## 312                           with_prints
## 313                      with_blastprodom
## 314                      with_scanprosite
## 315                           with_pfscan
## 316                              with_seg
## 317                             with_sfld
## 318                          with_signalp
## 319                            with_smart
## 320                      with_superfamily
## 321                          with_tigrfam
## 322                            with_tmhmm
## 323                              interpro
## 324                                family
## 325                                gene3d
## 326                                 hamap
## 327                            hmmpanther
## 328                                  pfam
## 329                                 pirsf
## 330                                prints
## 331                           blastprodom
## 332                           scanprosite
## 333                                pfscan
## 334                                 smart
## 335                           superfamily
## 336                               tigrfam
## 337            germ_line_variation_source
## 338              somatic_variation_source
## 339                    with_validated_snp
## 340                   so_mini_parent_name
##                                                          description
## 1                                           Chromosome/scaffold name
## 2                                                              Start
## 3                                                                End
## 4                                                         Band Start
## 5                                                           Band End
## 6                                                       Marker Start
## 7                                                         Marker End
## 8                                                      Encode region
## 9                                                             Strand
## 10                            e.g. 1:100:10000:-1, 1:100000:200000:1
## 11                                                   With CCDS ID(s)
## 12                                                 With ChEMBL ID(s)
## 13                             With Clone-based (Ensembl) gene ID(s)
## 14                       With Clone-based (Ensembl) transcript ID(s)
## 15                   With DataBase of Aberrant 3' Splice Sites ID(s)
## 16                   With DataBase of Aberrant 5' Splice Sites ID(s)
## 17                               With Ensembl Human Transcript ID(s)
## 18                              With Ensembl Human Translation ID(s)
## 19                             With EntrezGene transcript name ID(s)
## 20                            With European Nucleotide Archive ID(s)
## 21                                       With Expression Atlas ID(s)
## 22                                                 With GeneDB ID(s)
## 23                                                     With GO ID(s)
## 24                                             With GOSlim GOA ID(s)
## 25                                     With Havana translation ID(s)
## 26                                            With HGNC Symbol ID(s)
## 27                                   With HGNC transcript name ID(s)
## 28                                    With Human Protein Atlas ID(s)
## 29                                       With INSDC protein ID ID(s)
## 30                                With KEGG Pathway and Enzyme ID(s)
## 31                            With LRG display in Ensembl gene ID(s)
## 32                      With LRG display in Ensembl transcript ID(s)
## 33                        With MEROPS - the Peptidase Database ID(s)
## 34                                               With MIM gene ID(s)
## 35                                             With MIM morbid ID(s)
## 36                                                With miRBase ID(s)
## 37                                With miRBase transcript name ID(s)
## 38                                              With NCBI gene ID(s)
## 39                                                    With PDB ID(s)
## 40                                               With Reactome ID(s)
## 41                                          With Reactome gene ID(s)
## 42                                    With Reactome transcript ID(s)
## 43                                            With RefSeq mRNA ID(s)
## 44                                  With RefSeq mRNA predicted ID(s)
## 45                                           With RefSeq ncRNA ID(s)
## 46                                 With RefSeq ncRNA predicted ID(s)
## 47                                         With RefSeq peptide ID(s)
## 48                               With RefSeq peptide predicted ID(s)
## 49                                                   With RFAM ID(s)
## 50                                   With RFAM transcript name ID(s)
## 51                                             With RNAcentral ID(s)
## 52                                         With UCSC Stable ID ID(s)
## 53                                                With UniParc ID(s)
## 54                                    With UniProtKB Gene Name ID(s)
## 55                                   With UniProtKB/Swiss-Prot ID(s)
## 56                                       With UniProtKB/TrEMBL ID(s)
## 57                                       With Vega translation ID(s)
## 58                                               With WikiGene ID(s)
## 59                          Gene stable ID(s) [e.g. ENSG00000000003]
## 60          Gene stable ID(s) with version [e.g. ENSG00000000003.14]
## 61                    Transcript stable ID(s) [e.g. ENST00000000233]
## 62     Transcript stable ID(s) with version [e.g. ENST00000000233.9]
## 63                       Protein stable ID(s) [e.g. ENSP00000000233]
## 64        Protein stable ID(s) with version [e.g. ENSP00000000233.5]
## 65                                 Exon ID(s) [e.g. ENSE00000327880]
## 66                                       Gene Name(s) [e.g. RF00100]
## 67                          Transcript Name(s) [e.g. RF00100.12-201]
## 68                                          CCDS ID(s) [e.g. CCDS10]
## 69                                 ChEMBL ID(s) [e.g. CHEMBL1075092]
## 70                Clone-based (Ensembl) gene ID(s) [e.g. AB015752.1]
## 71      Clone-based (Ensembl) transcript ID(s) [e.g. AB015752.1-201]
## 72         DataBase of Aberrant 3' Splice Sites name(s) [e.g. ABCC8]
## 73               DataBase of Aberrant 3' Splice Sites ID(s) [e.g. 1]
## 74         DataBase of Aberrant 5' Splice Sites name(s) [e.g. ABCD1]
## 75               DataBase of Aberrant 5' Splice Sites ID(s) [e.g. 1]
## 76             Ensembl Human Transcript ID(s) [e.g. ENST00000075430]
## 77            Ensembl Human Translation ID(s) [e.g. ENSP00000075430]
## 78                 EntrezGene transcript name ID(s) [e.g. ALDOA-222]
## 79                   European Nucleotide Archive ID(s) [e.g. A06800]
## 80                     Expression Atlas ID(s) [e.g. ENSG00000000003]
## 81                              GeneDB ID(s) [e.g. Smp_009580.1:pep]
## 82                                        GO ID(s) [e.g. GO:0000002]
## 83                                GOSlim GOA ID(s) [e.g. GO:0000003]
## 84                      Havana translation ID(s) [e.g. P00000281639]
## 85                                        HGNC ID(s) [e.g. HGNC:100]
## 86                                        HGNC symbol(s) [e.g. A1BG]
## 87                        HGNC transcript name ID(s) [e.g. A1BG-201]
## 88                                Human Protein Atlas ID(s) [e.g. 1]
## 89                               INSDC protein ID(s) [e.g. AAA02487]
## 90                KEGG Pathway and Enzyme ID(s) [e.g. 00010+1.1.1.1]
## 91                    LRG display in Ensembl gene ID(s) [e.g. LRG_1]
## 92         LRG display in Ensembl transcript ID(s) [e.g. LRG_1008t1]
## 93              MEROPS - the Peptidase Database ID(s) [e.g. A01.001]
## 94                               MIM gene accession(s) [e.g. 100640]
## 95                             MIM morbid accession(s) [e.g. 100100]
## 96                             miRBase accession(s) [e.g. MI0000060]
## 97                                 miRBase ID(s) [e.g. hsa-let-7a-1]
## 98           miRBase transcript name ID(s) [e.g. hsa-mir-1253.1-201]
## 99                                          NCBI gene ID(s) [e.g. 1]
## 100                                            PDB ID(s) [e.g. 10GS]
## 101                              Reactome ID(s) [e.g. R-HSA-1059683]
## 102                         Reactome gene ID(s) [e.g. R-HSA-1059683]
## 103                   Reactome transcript ID(s) [e.g. R-HSA-1257604]
## 104                               RefSeq mRNA ID(s) [e.g. NM_000014]
## 105                  RefSeq mRNA predicted ID(s) [e.g. XM_003118719]
## 106                              RefSeq ncRNA ID(s) [e.g. NR_000005]
## 107                 RefSeq ncRNA predicted ID(s) [e.g. XR_001731188]
## 108                            RefSeq peptide ID(s) [e.g. NP_000005]
## 109               RefSeq peptide predicted ID(s) [e.g. XP_003118767]
## 110                                        RFAM ID(s) [e.g. RF00001]
## 111                  RFAM transcript name ID(s) [e.g. RF00001.1-201]
## 112                            RNAcentral ID(s) [e.g. UPI0001765128]
## 113                              UCSC Stable ID(s) [e.g. uc001aak.4]
## 114                               UniParc ID(s) [e.g. UPI000000003C]
## 115                      UniProtKB Gene Name ID(s) [e.g. A0A023T6R1]
## 116                     UniProtKB/Swiss-Prot ID(s) [e.g. A0A075B6H9]
## 117                         UniProtKB/TrEMBL ID(s) [e.g. A0A023T6R1]
## 118                             Vega translation ID(s) [e.g. 134942]
## 119                                     WikiGene name(s) [e.g. A1BG]
## 120                                          WikiGene ID(s) [e.g. 1]
## 121                                    With AFFY HC G110 probe ID(s)
## 122                                   With AFFY HG Focus probe ID(s)
## 123                                   With AFFY HG U133A probe ID(s)
## 124                                 With AFFY HG U133A 2 probe ID(s)
## 125                                   With AFFY HG U133B probe ID(s)
## 126                             With AFFY HG U133 Plus 2 probe ID(s)
## 127                                    With AFFY HG U95A probe ID(s)
## 128                                  With AFFY HG U95Av2 probe ID(s)
## 129                                    With AFFY HG U95B probe ID(s)
## 130                                    With AFFY HG U95C probe ID(s)
## 131                                    With AFFY HG U95D probe ID(s)
## 132                                    With AFFY HG U95E probe ID(s)
## 133                                    With AFFY HTA 2 0 probe ID(s)
## 134                             With AFFY HuEx 1 0 st v2 probe ID(s)
## 135                                   With AFFY HuGeneFL probe ID(s)
## 136                           With AFFY HuGene 1 0 st v1 probe ID(s)
## 137                           With AFFY HuGene 2 0 st v1 probe ID(s)
## 138                                  With AFFY PrimeView probe ID(s)
## 139                                   With AFFY U133 X3P probe ID(s)
## 140                                 With AGILENT CGH 44b probe ID(s)
## 141                                 With AGILENT GPL6848 probe ID(s)
## 142                   With AGILENT SurePrint G3 GE 8x60k probe ID(s)
## 143                With AGILENT SurePrint G3 GE 8x60k v2 probe ID(s)
## 144                             With AGILENT WholeGenome probe ID(s)
## 145                    With AGILENT WholeGenome 4x44k v1 probe ID(s)
## 146                    With AGILENT WholeGenome 4x44k v2 probe ID(s)
## 147                               With CODELINK CODELINK probe ID(s)
## 148                          With ILLUMINA HumanHT 12 V3 probe ID(s)
## 149                          With ILLUMINA HumanHT 12 V4 probe ID(s)
## 150                          With ILLUMINA HumanRef 8 V3 probe ID(s)
## 151                           With ILLUMINA HumanWG 6 V1 probe ID(s)
## 152                           With ILLUMINA HumanWG 6 V2 probe ID(s)
## 153                           With ILLUMINA HumanWG 6 V3 probe ID(s)
## 154                                With PHALANX OneArray probe ID(s)
## 155                          AFFY HC G110 probe ID(s) [e.g. 1393_at]
## 156                     AFFY HG Focus probe ID(s) [e.g. 203953_s_at]
## 157                       AFFY HG U133A probe ID(s) [e.g. 211600_at]
## 158                     AFFY HG U133A 2 probe ID(s) [e.g. 211600_at]
## 159                     AFFY HG U133B probe ID(s) [e.g. 224373_s_at]
## 160              AFFY HG U133 Plus 2 probe ID(s) [e.g. 1553551_s_at]
## 161                         AFFY HG U95A probe ID(s) [e.g. 33904_at]
## 162                       AFFY HG U95Av2 probe ID(s) [e.g. 33904_at]
## 163                         AFFY HG U95B probe ID(s) [e.g. 57739_at]
## 164                         AFFY HG U95C probe ID(s) [e.g. 60059_at]
## 165                         AFFY HG U95D probe ID(s) [e.g. 81914_at]
## 166                         AFFY HG U95E probe ID(s) [e.g. 88289_at]
## 167                    AFFY HTA 2 0 probe ID(s) [e.g. TC0M000002.hg]
## 168                   AFFY HuEx 1 0 st v2 probe ID(s) [e.g. 4037578]
## 169                       AFFY HuGeneFL probe ID(s) [e.g. Z70759_at]
## 170                 AFFY HuGene 1 0 st v1 probe ID(s) [e.g. 8165644]
## 171                AFFY HuGene 2 0 st v1 probe ID(s) [e.g. 17100641]
## 172                  AFFY PrimeView probe ID(s) [e.g. 11761516_x_at]
## 173                 AFFY U133 X3P probe ID(s) [e.g. 1553551_3p_s_at]
## 174                  AGILENT CGH 44b probe ID(s) [e.g. A_14_P121209]
## 175                  AGILENT GPL6848 probe ID(s) [e.g. A_24_P179339]
## 176    AGILENT SurePrint G3 GE 8x60k probe ID(s) [e.g. A_24_P182122]
## 177 AGILENT SurePrint G3 GE 8x60k v2 probe ID(s) [e.g. A_24_P182122]
## 178               AGILENT WholeGenome probe ID(s) [e.g. A_24_P42453]
## 179     AGILENT WholeGenome 4x44k v1 probe ID(s) [e.g. A_24_P179339]
## 180     AGILENT WholeGenome 4x44k v2 probe ID(s) [e.g. A_24_P182122]
## 181                    CODELINK CODELINK probe ID(s) [e.g. GE555137]
## 182           ILLUMINA HumanHT 12 V3 probe ID(s) [e.g. ILMN_1721028]
## 183           ILLUMINA HumanHT 12 V4 probe ID(s) [e.g. ILMN_3308961]
## 184           ILLUMINA HumanRef 8 V3 probe ID(s) [e.g. ILMN_1804758]
## 185              ILLUMINA HumanWG 6 V1 probe ID(s) [e.g. 0005910053]
## 186            ILLUMINA HumanWG 6 V2 probe ID(s) [e.g. ILMN_1721028]
## 187            ILLUMINA HumanWG 6 V3 probe ID(s) [e.g. ILMN_1721028]
## 188                PHALANX OneArray probe ID(s) [e.g. PH_hs_0036541]
## 189                                              Transcript count &gt;=
## 190                                              Transcript count &lt;=
## 191                                                             Type
## 192                                                  Transcript Type
## 193                                                    Source (gene)
## 194                                              Source (transcript)
## 195                                   Transcript Support Level (TSL)
## 196                                         GENCODE basic annotation
## 197                                                APPRIS annotation
## 198                                            Phenotype description
## 199                                                 Phenotype source
## 200                                            Parent term accession
## 201                                                 Parent term name
## 202                                                 GO Evidence code
## 203                                           Paralogous Human Genes
## 204                                 Orthologous Algerian mouse Genes
## 205                                         Orthologous Alpaca Genes
## 206                                   Orthologous Amazon molly Genes
## 207                                 Orthologous Angola colobus Genes
## 208                                   Orthologous Anole lizard Genes
## 209                                      Orthologous Armadillo Genes
## 210                        Orthologous Black snub-nosed monkey Genes
## 211                       Orthologous Bolivian squirrel monkey Genes
## 212                                         Orthologous Bonobo Genes
## 213                           Orthologous Brazilian guinea pig Genes
## 214                                       Orthologous Bushbaby Genes
## 215                                 Orthologous C.intestinalis Genes
## 216                                     Orthologous C.savignyi Genes
## 217                         Orthologous Caenorhabditis elegans Genes
## 218                                       Orthologous Capuchin Genes
## 219                                            Orthologous Cat Genes
## 220                                      Orthologous Cave fish Genes
## 221                                        Orthologous Chicken Genes
## 222                                     Orthologous Chimpanzee Genes
## 223                        Orthologous Chinese hamster CHOK1GS Genes
## 224                         Orthologous Chinese hamster CriGri Genes
## 225                       Orthologous Chinese softshell turtle Genes
## 226                                            Orthologous Cod Genes
## 227                                     Orthologous Coelacanth Genes
## 228                              Orthologous Coquerel's sifaka Genes
## 229                                            Orthologous Cow Genes
## 230                            Orthologous Crab-eating macaque Genes
## 231                                Orthologous Damara mole rat Genes
## 232                                           Orthologous Degu Genes
## 233                                            Orthologous Dog Genes
## 234                                        Orthologous Dolphin Genes
## 235                                          Orthologous Drill Genes
## 236                                           Orthologous Duck Genes
## 237                                       Orthologous Elephant Genes
## 238                                         Orthologous Ferret Genes
## 239                                     Orthologous Flycatcher Genes
## 240                                       Orthologous Fruitfly Genes
## 241                                           Orthologous Fugu Genes
## 242                                         Orthologous Gibbon Genes
## 243                                           Orthologous Goat Genes
## 244                                 Orthologous Golden Hamster Genes
## 245                       Orthologous Golden snub-nosed monkey Genes
## 246                                        Orthologous Gorilla Genes
## 247                                     Orthologous Guinea Pig Genes
## 248                                       Orthologous Hedgehog Genes
## 249                                          Orthologous Horse Genes
## 250                                          Orthologous Hyrax Genes
## 251                                   Orthologous Kangaroo rat Genes
## 252                                        Orthologous Lamprey Genes
## 253                         Orthologous Lesser Egyptian jerboa Genes
## 254                         Orthologous Lesser hedgehog tenrec Genes
## 255                         Orthologous Long-tailed chinchilla Genes
## 256                              Orthologous Ma's night monkey Genes
## 257                                        Orthologous Macaque Genes
## 258                                       Orthologous Marmoset Genes
## 259                                         Orthologous Medaka Genes
## 260                                        Orthologous Megabat Genes
## 261                                       Orthologous Microbat Genes
## 262                                          Orthologous Mouse Genes
## 263                                    Orthologous Mouse Lemur Genes
## 264                          Orthologous Naked mole-rat female Genes
## 265                            Orthologous Naked mole-rat male Genes
## 266                   Orthologous Northern American deer mouse Genes
## 267                                   Orthologous Olive baboon Genes
## 268                                        Orthologous Opossum Genes
## 269                                      Orthologous Orangutan Genes
## 270                                          Orthologous Panda Genes
## 271                                            Orthologous Pig Genes
## 272                             Orthologous Pig-tailed macaque Genes
## 273                                           Orthologous Pika Genes
## 274                                      Orthologous Platyfish Genes
## 275                                       Orthologous Platypus Genes
## 276                                   Orthologous Prairie vole Genes
## 277                                         Orthologous Rabbit Genes
## 278                                            Orthologous Rat Genes
## 279                                   Orthologous Ryukyu mouse Genes
## 280                       Orthologous Saccharomyces cerevisiae Genes
## 281                                          Orthologous Sheep Genes
## 282                                          Orthologous Shrew Genes
## 283                                    Orthologous Shrew mouse Genes
## 284                                          Orthologous Sloth Genes
## 285                                 Orthologous Sooty mangabey Genes
## 286                                    Orthologous Spotted gar Genes
## 287                                       Orthologous Squirrel Genes
## 288                                    Orthologous Stickleback Genes
## 289                                        Orthologous Tarsier Genes
## 290                                Orthologous Tasmanian devil Genes
## 291                                      Orthologous Tetraodon Genes
## 292                                        Orthologous Tilapia Genes
## 293                                     Orthologous Tree Shrew Genes
## 294                                         Orthologous Turkey Genes
## 295         Orthologous Upper Galilee mountains blind mole rat Genes
## 296                                     Orthologous Vervet-AGM Genes
## 297                                        Orthologous Wallaby Genes
## 298                                        Orthologous Xenopus Genes
## 299                                    Orthologous Zebra Finch Genes
## 300                                      Orthologous Zebrafish Genes
## 301                                              With Interpro ID(s)
## 302                                With Ensembl Protein Family ID(s)
## 303                                           With PDB-ENSP mappings
## 304                                                         With CDD
## 305                                                With Gene3D ID(s)
## 306                                                 With HAMAP ID(s)
## 307                                                  With MobiDBLite
## 308                                       With Coiled-coils (Ncoils)
## 309                                            With hmmpanther ID(s)
## 310                                           With Pfam domain ID(s)
## 311                                          With PIRSF domain ID(s)
## 312                                         With Prints domain ID(s)
## 313                                           With blastprodom ID(s)
## 314                                      With PROSITE patterns ID(s)
## 315                                      With PROSITE profiles ID(s)
## 316                                        With Low complexity (Seg)
## 317                                                        With SFLD
## 318                                     With Cleavage site (Signalp)
## 319                                         With SMART domains ID(s)
## 320                                   With Superfamily domains ID(s)
## 321                                        With TIGRFAM domain ID(s)
## 322                                       With Transmembrane helices
## 323                                  Interpro ID(s) [e.g. IPR000001]
## 324                Ensembl Protein Family ID(s) [e.g. PTHR10000_SF7]
## 325                                   Gene3D ID(s) [e.g. 1.10.10.10]
## 326                                      HAMAP ID(s) [e.g. MF_00001]
## 327                                hmmpanther ID(s) [e.g. PTHR10003]
## 328                                 Pfam domain ID(s) [e.g. PF00001]
## 329                            PIRSF domain ID(s) [e.g. PIRSF000019]
## 330                               Prints domain ID(s) [e.g. PR00001]
## 331                                blastprodom ID(s) [e.g. PD000328]
## 332                            PROSITE patterns ID(s) [e.g. PS00010]
## 333                            PROSITE profiles ID(s) [e.g. PS01031]
## 334                               SMART domains ID(s) [e.g. SM00002]
## 335                       Superfamily domains ID(s) [e.g. SSF100879]
## 336                            TIGRFAM domain ID(s) [e.g. TIGR00002]
## 337                limit to genes with germline variant data sources
## 338                      limit to genes with somatic variant sources
## 339                                      Variant supporting evidence
## 340                                                 Parent term name
```
---
##Bringing in External data. Gene models through Biomart 

 
 
Here we select only genes which have been annotated by both havana and ensembl (so called *Golden Transcripts*) 
 

```r
bgrTrack &lt;- BiomartGeneRegionTrack(genome="hg19", 
                                   start=26591341, 
                                   end=27034958, 
                                   chromosome = "chr7", 
                                   name="ENSEMBL",               
                                  filter=list(source="ensembl_havana")) 
```
---
##Bringing in External data. Gene models through Biomart 

 
 
Once we have retrieved our filtered gene models we can plot them as before. 
 

```r
plotTracks(bgrTrack) 
```

![](Viz_part1_files/figure-html/e-1.png)&lt;!-- --&gt;
---
##Bringing in External data. Tracks from UCSC 

 
 
A well known browser and source of genomic data and annotation is the [UCSC genome browser](https://genome.ucsc.edu/). **Gviz** can create track directly from UCSC tables using the functionality from **rtracklayer** Bioconductor package. 
 
The **Ucsctrack()** constructor and object allow for the query and track construction of a variety of data types. The **Ucsctrack()** function therefore requires us to specify the track type we expect using the **trackType** parameter as well as the required UCSC table using the **track** parameter.  
 
---
##Bringing in External data. Tracks from UCSC 

 
 
To understand which tables are available we can query the **rtracktables** package to identify track and table names. 
 

```r
library(rtracklayer) 
session &lt;- browserSession() 
genome(session) &lt;- "hg19" 
trackNames(session) 
```

```
##               Base Position              Alt Haplotypes 
##                     "ruler"              "altLocations" 
##                    Assembly               BAC End Pairs 
##                      "gold"               "bacEndPairs" 
##                   BU ORChID             Chromosome Band 
##          "wgEncodeBuOrchid"                  "cytoBand" 
##               deCODE Recomb                ENCODE Pilot 
##                "decodeRmap"             "encodeRegions" 
##                 FISH Clones            Fosmid End Pairs 
##                "fishClones"               "fosEndPairs" 
##                         Gap                  GC Percent 
##                       "gap"                   "gc5Base" 
##                GRC Incident             GRC Map Contigs 
##             "grcIncidentDb"                   "ctgPos2" 
##           GRC Patch Release                   Hg18 Diff 
##         "altSeqComposite10"            "hg19ContigDiff" 
##                   Hg38 Diff                Hi Seq Depth 
##            "hg38ContigDiff"                "hiSeqDepth" 
##                       INSDC                 LRG Regions 
##               "ucscToINSDC"                       "lrg" 
##                 Map Contigs                 Mappability 
##                    "ctgPos"        "wgEncodeMapability" 
##                 Recomb Rate                  RefSeq Acc 
##                "recombRate"              "ucscToRefSeq" 
##               Restr Enzymes                 Short Match 
##                   "cutters"                "oligoMatch" 
##                 STS Markers                  Wiki Track 
##                    "stsMap"                 "wikiTrack" 
##                  UCSC Genes                 NCBI RefSeq 
##                 "knownGene"           "refSeqComposite" 
##               AceView Genes                    AUGUSTUS 
##                   "acembly"              "augustusGene" 
##                        CCDS                   CRISPR... 
##                  "ccdsGene"                    "crispr" 
##               Ensembl Genes                     EvoFold 
##                   "ensGene"                   "evofold" 
##                    Exoniphy                  GENCODE... 
##                  "exoniphy"      "wgEncodeGencodeSuper" 
##                Geneid Genes               Genscan Genes 
##                    "geneid"                   "genscan" 
##                   H-Inv 7.0           IKMC Genes Mapped 
##           "hinv70Composite"                    "hgIkmc" 
##                 lincRNAs...             LRG Transcripts 
##                  "lincRNAs"          "lrgTranscriptAli" 
##                   MGC Genes                      N-SCAN 
##               "mgcFullMrna"                 "nscanGene" 
##              Old UCSC Genes              ORFeome Clones 
##             "knownGeneOld6"               "orfeomeMrna" 
##                Other RefSeq           Pfam in UCSC Gene 
##               "xenoRefGene"              "ucscGenePfam" 
##            Retroposed Genes                   SGP Genes 
##             "ucscRetroAli5"                   "sgpGene" 
##                   SIB Genes                   sno/miRNA 
##                   "sibGene"                     "wgRna" 
##                 TransMap...                  tRNA Genes 
##                "transMapV4"                     "tRNAs" 
##             UCSC Alt Events                     UniProt 
##                  "knownAlt"                   "uniprot" 
##                  Vega Genes               Yale Pseudo60 
##         "vegaGeneComposite"              "pseudoYale60" 
##                Publications                ClinGen CNVs 
##                      "pubs"             "iscaComposite" 
##            ClinVar Variants                Coriell CNVs 
##                   "clinvar"             "coriellDelDup" 
##              COSMIC Regions               DECIPHER CNVs 
##             "cosmicRegions"                  "decipher" 
##           Development Delay                    GAD View 
##               "cnvDevDelay"                       "gad" 
##           Gene Interactions                 GeneReviews 
##              "interactions"               "geneReviews" 
##                GWAS Catalog               HGMD Variants 
##               "gwasCatalog"                      "hgmd" 
##                Lens Patents               LOVD Variants 
##                    "patSeq"                      "lovd" 
##               MGI Mouse QTL                OMIM Alleles 
##              "jaxQtlMapped"                 "omimAvSnp" 
##                  OMIM Genes             OMIM Pheno Loci 
##                 "omimGene2"              "omimLocation" 
##               RGD Human QTL                 RGD Rat QTL 
##                    "rgdQtl"                 "rgdRatQtl" 
##                     SNPedia            UniProt Variants 
##                   "snpedia"                     "spMut" 
##               Web Sequences                   CGAP SAGE 
##              "pubsBingBlat"                  "cgapSage" 
##                 Gene Bounds                       H-Inv 
##                "rnaCluster"              "HInvGeneMrna" 
##                  Human ESTs                 Human mRNAs 
##                       "est"                      "mrna" 
##           Human RNA Editing                  Other ESTs 
##                    "darned"                   "xenoEst" 
##                 Other mRNAs                     Poly(A) 
##                  "xenoMrna"                     "polyA" 
##                   PolyA-Seq            SIB Alt-Splicing 
##             "polyASeqSites"                "sibTxGraph" 
##                Spliced ESTs                     UniGene 
##                 "intronEst"                 "uniGene_3" 
##                   GTEx Gene             GTEx Transcript 
##                  "gtexGene"            "gtexTranscExpr" 
##             Affy Exon Array                  Affy GNF1H 
##             "affyExonArray"                 "affyGnf1h" 
##                Affy RNA Loc                    Affy U95 
##       "wgEncodeAffyRnaChip"                   "affyU95" 
##                   Affy U133              Affy U133Plus2 
##                  "affyU133"             "affyU133Plus2" 
##                 Allen Brain               Burge RNA-seq 
##             "allenBrainAli" "burgeRnaSeqGemMapperAlign" 
##          CSHL Small RNA-seq           ENC Exon Array... 
##   "wgEncodeCshlShortRnaSeq"    "wgEncodeExonArraySuper" 
##             ENC ProtGeno...              ENC RNA-seq... 
##     "wgEncodeProtGenoSuper"       "wgEncodeRnaSeqSuper" 
##                 GIS RNA PET                 GNF Atlas 2 
##         "wgEncodeGisRnaPet"                 "gnfAtlas2" 
##           GWIPS-viz Riboseq               Illumina WG-6 
##           "gwipsvizRiboseq"            "illuminaProbes" 
##                PeptideAtlas                qPCR Primers 
##          "peptideAtlas2014"               "qPcrPrimers" 
##              RIKEN CAGE Loc                Sestan Brain 
##         "wgEncodeRikenCage"          "sestanBrainAtlas" 
##        ENCODE Regulation...          GTEx Combined eQTL 
##               "wgEncodeReg"           "gtexEqtlCluster" 
##            GTEx Tissue eQTL                 CD34 DnaseI 
##            "gtexEqtlTissue"                "eioJcviNAS" 
##              CpG Islands...            ENC Chromatin... 
##            "cpgIslandSuper"        "wgEncodeChromSuper" 
##           ENC DNA Methyl...          ENC DNase/FAIRE... 
##    "wgEncodeDnaMethylSuper"        "wgEncodeDNAseSuper" 
##              ENC Histone...          ENC RNA Binding... 
##      "wgEncodeHistoneSuper"          "wgEncodeRbpSuper" 
##           ENC TF Binding...              FSU Repli-chip 
##    "wgEncodeTfBindingSuper"      "wgEncodeFsuRepliChip" 
##             Genome Segments           NKI Nuc Lamina... 
##   "wgEncodeAwgSegmentation"              "laminB1Super" 
##                    ORegAnno            Stanf Nucleosome 
##                  "oreganno"         "wgEncodeSydhNsome" 
##             SUNY SwitchGear              SwitchGear TSS 
##    "wgEncodeSunySwitchgear"               "switchDbTss" 
##              TFBS Conserved              TS miRNA sites 
##             "tfbsConsSites"               "targetScanS" 
##           UCSF Brain Methyl             UMMS Brain Hist 
##           "ucsfBrainMethyl"         "uMassBrainHistone" 
##                UW Repli-seq             Vista Enhancers 
##        "wgEncodeUwRepliSeq"            "vistaEnhancers" 
##                Conservation                 Cons 46-Way 
##                "cons100way"                 "cons46way" 
##            Cons Indels MmCf                     Evo Cpg 
##      "consIndelsHgMmCanFam"                    "evoCpg" 
##                        GERP              phastBias gBGC 
##              "allHg19RS_BW"                 "phastBias" 
##           Primate Chain/Net         Placental Chain/Net 
##           "primateChainNet"         "placentalChainNet" 
##        Vertebrate Chain/Net                 5% Lowest S 
##        "vertebrateChainNet"                "ntSssTop5p" 
##            H-C Coding Diffs           Neandertal Methyl 
##      "ntHumChimpCodingDiff"     "neandertalMethylation" 
##              Neandertal Seq                      S SNPs 
##                "ntSeqReads"                 "ntSssSnps" 
##            Sel Swp Scan (S)             Denisova Methyl 
##          "ntSssZScorePMVar"       "denisovaMethylation" 
##                Denisova Seq           Denisova Variants 
##            "dhcBamDenisova"       "dhcVcfDenisovaPinky" 
##            Mod Hum Variants              Modern Derived 
##              "dhcVcfModern"           "dhcHumDerDenAnc" 
##            Common SNPs(150)            Common SNPs(147) 
##              "snp150Common"              "snp147Common" 
##            Common SNPs(146)            Common SNPs(144) 
##              "snp146Common"              "snp144Common" 
##            Common SNPs(142)            Common SNPs(141) 
##              "snp142Common"              "snp141Common" 
##               All SNPs(150)               All SNPs(147) 
##                    "snp150"                    "snp147" 
##               All SNPs(146)               All SNPs(144) 
##                    "snp146"                    "snp144" 
##               All SNPs(142)               All SNPs(141) 
##                    "snp142"                    "snp141" 
##           Flagged SNPs(150)           Flagged SNPs(147) 
##             "snp150Flagged"             "snp147Flagged" 
##           Flagged SNPs(146)           Flagged SNPs(144) 
##             "snp146Flagged"             "snp144Flagged" 
##           Flagged SNPs(142)           Flagged SNPs(141) 
##             "snp142Flagged"             "snp141Flagged" 
##             Mult. SNPs(150)             Mult. SNPs(147) 
##                "snp150Mult"                "snp147Mult" 
##             Mult. SNPs(146)             Mult. SNPs(144) 
##                "snp146Mult"                "snp144Mult" 
##             Mult. SNPs(142)             Mult. SNPs(138) 
##                "snp142Mult"                "snp138Mult" 
##               All SNPs(138)            Common SNPs(138) 
##                    "snp138"              "snp138Common" 
##           Flagged SNPs(138)            1000G Ph1 Accsbl 
##             "snp138Flagged"    "tgpPhase1Accessibility" 
##              1000G Ph1 Vars            1000G Ph3 Accsbl 
##                 "tgpPhase1"    "tgpPhase3Accessibility" 
##              1000G Ph3 Vars              DGV Struct Var 
##                 "tgpPhase3"                   "dgvPlus" 
##                EVS Variants                        ExAC 
##                "evsEsp6500"                      "exac" 
##             Genome Variants                 GIS DNA PET 
##                     "pgSnp"         "wgEncodeGisDnaPet" 
##               HAIB Genotype                 HapMap SNPs 
##      "wgEncodeHaibGenotype"                "hapmapSnps" 
##            HGDP Allele Freq              SNP/CNV Arrays 
##                   "hgdpGeo"            "genotypeArrays" 
##                RepeatMasker            Interrupted Rpts 
##                      "rmsk"             "nestedRepeats" 
##              Microsatellite              NumtS Sequence 
##                  "microsat"                   "numtSeq" 
##              Segmental Dups                  Self Chain 
##          "genomicSuperDups"                 "chainSelf" 
##              Simple Repeats 
##              "simpleRepeat"
```

```r
query &lt;- ucscTableQuery(session, "Ensembl Genes", 
                        GRangesForUCSCGenome("hg19", "chr7", 
                                             IRanges(26591341,27034958))) 
tableNames(query) 
```

```
## [1] "ensGene"           "ccdsInfo"          "ensGtp"           
## [4] "ensPep"            "ensemblSource"     "ensemblToGeneName"
## [7] "knownToEnsembl"
```
---
##Bringing in External data. Tracks from UCSC 

 
 

```r
query &lt;- ucscTableQuery(session, "Ensembl Genes", 
                        GRangesForUCSCGenome("hg19", "chr7", 
                                             IRanges(26591341,27034958))) 
tableNames(query) 
```

```
## [1] "ensGene"           "ccdsInfo"          "ensGtp"           
## [4] "ensPep"            "ensemblSource"     "ensemblToGeneName"
## [7] "knownToEnsembl"
```
---
##Bringing in External data. Tracks from UCSC 

 
 
 

```r
ucscTrack &lt;- UcscTrack(genome = "hg19", 
                       chromosome = "chr7", 
                       track = "ensGene", 
                       from = 26591341, 
                       to = 27034958, 
                       trackType = "GeneRegionTrack", 
                       rstarts = "exonStarts", 
                       rends = "exonEnds", 
                       gene ="name", 
                       symbol = "name2", 
                       transcript = "name", 
                       strand = "strand" 
) 
```
 

To build the UCSC annotation as a **GeneRegionTrack** we must specify some information specific to **GeneRegionTrack** objects. This includes the "rstarts" and "rends". You can consult the help for **GeneRegionTrack()** (*?GeneRegionTrack to see from in R*) to see full parameters required for **UcscTrack** objects. 
---
##Bringing in External data. Tracks from UCSC 

 
 
Now we can compare the Ensembl gene builds from the two different sources.  
 
Notable differences in the annotation include the absense of some transcipts due to the additional filter applied in our **BiomartGeneRegionTrack** object creation. 
 

```r
plotTracks(c(bgrTrack,ucscTrack), 
           from = 26591341,to = 27034958) 
```

![](Viz_part1_files/figure-html/e1-1.png)&lt;!-- --&gt;
 
---
##Bringing in External data. Tracks from UCSC as DataTrack 

 
 
By the same method we can take advantage of other types of UCSC data. 
 
In this example we capture the Conservation in the phyloP100wayAll table over and around our previously investigated ChIP-seq reads peak. 
 
Here we specify the data to be returned as a **DataTrack** object and the display type to be "hist". Here we are creating a **DataTrack** so would consult **DataTrack()** help *(?DataTrack)* to get full parameter list. 
 

```r
conservationTrack &lt;- UcscTrack(genome = "hg19", chromosome = "chr5",track = "Conservation", table = "phyloP100wayAll",from = 135313003, to = 135313570, trackType = "DataTrack",start = "start", end = "end", data = "score",type = "hist", window = "auto", col.histogram = "darkblue",fill.histogram = "darkblue", ylim = c(-3.7, 4),name = "Conservation") 
```
 
---
##Bringing in External data. Tracks from UCSC as DataTrack 

 
 
With the inclusion of conservation alongside the coverage from CTCF peaks we can see a spike in conservation around the CTCF peak summit. We include a relative scale and increase the size of text for completeness. 
 

 

 

```r
plotTracks(c(conservationTrack,peakReads,genomeAxis), 
           from=135313003, 
           to=135313570, 
           chromosome = "chr5", 
           type = c("hist","coverage"), 
           sizes = c(1,1,0.2), 
           cex=2) 
```

![](Viz_part1_files/figure-html/i1-1.png)&lt;!-- --&gt;
 
---
##Exercises 

 
 
 
Time for exercises! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/exercises/ExternalData_Exercises.html) 
 
---
##Solutions 

 
 
 
Time for solutions! [Link here](https://mrccsc.github.io/VisualisingGenomicsData/solutions/ExternalData_Solutions.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
